Sprint 06

zebolovnik opened the pull request:
* Добавлена работа с сетью, список фильмов загружается из API IMDb.
* Обработаны состояния приложения при загрузке фильмов и получения ошибок.

zebolovnik committed:
Добавлена обработка ошибок сети с функционалом повторной попытки

- Реализован метод  для отображения алерта при возникновении ошибки сети.
- Добавлена кнопка повторной попытки в алерт для повторной загрузки данных.
- Интегрирована скрытие и отображение индикатора загрузки в зависимости от статуса ошибки сети.

zebolovnik committed:
Добавлены структуры MostPopularMovies и MoviesLoader для загрузки фильмов через NetworkClient

zebolovnik committed:
Добавлен метод simulateError(); Подготовка к ревью sprint_06

andrew-kuzn-yndx commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -20,25 +20,31 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
     @IBOutlet private var counterLabel: UILabel!
     @IBOutlet private var noButton: UIButton!
```
Молодец, что указал private для аутлетов! Также хорошей практикой является указание модификатора weak к IBOutlet. Вероятность retain cycle тут мала, но и нам ничего не стоит написать слабые ссылки)

andrew-kuzn-yndx commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -47,10 +53,12 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    
     @IBAction private func yesButtonClicked(_ sender: UIButton) {
         setButtonsEnabled(false)
         guard let currentQuestion = currentQuestion else {
```
Блоки кода внутри IBAction. Лучше будет вывести эту логику в отдельные функции, и вызывать уже их из IBAction. Например:
```
    @IBAction func yesButtonClicked(_ sender: Any) {
        answerGived(answer: true)
    }
    
    private func answerGived(answer: Bool) {
        реализация
    }
```

Тем более код в yesButtonClicked и в noButtonClicked почти одинаковый, поэтому для них можно сделать универсальную функцию с аргументом на вход. Так будет правильнее

andrew-kuzn-yndx commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -71,13 +79,45 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    
+    private func hideLoadingIndicator() {
+        activityIndicator.stopAnimating()
+        activityIndicator.isHidden = true
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 
**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде. А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

andrew-kuzn-yndx commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -1,34 +1,83 @@
 import Foundation
 
 class QuestionFactory: QuestionFactoryProtocol {
```
Тут можно добавить final. Ключевое слово final в Swift позволяет объявить класс, свойство или метод, который не может быть переопределён. Это не только защищает твой код от случайного переопределения, но и помогает компилятору Swift производить дополнительные оптимизации кода, что улучшает производительность. 
Да и просто является хорошим тоном: при ревью сразу считывается намерение разработчика наследоваться от этого класса или нет

Рекомендую добавлять этот модификатор ко всем классам при их создании, а затем, если вдруг понадобится наследоваться от какого-то класса, то компилятор подскажет, что нужно убрать final)

andrew-kuzn-yndx commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -1,34 +1,83 @@
+//        QuizQuestion(image: "The Ice Age Adventures of Buck Wild", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false),
+//        QuizQuestion(image: "Tesla", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false),
+//        QuizQuestion(image: "Vivarium", text: "Рейтинг этого фильма больше чем 6?", correctAnswer: false)
+//    ]
```
Старайся не оставлять такое, правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код без лишних принтов и закомментированных блоков)

zebolovnik committed:
1. Добавлен модификатор weak для IBOutlet свойств, чтобы избежать потенциальных утечек памяти.
2. Вынесена общая логика из yesButtonClicked и noButtonClicked в универсальный метод answerGieed(answer:). Уменьшено дублирование кода.
3. Настроено свойство hidesWhenStopped у UIActivityIndicatorView для автоматического скрытия/отображения индикатора без ручного управления свойством isHidden.

zebolovnik committed:
class QuestionFactory таперь final
